document/js-api/index.bs

```diff
@@ -1329,6 +1383,133 @@ The algorithm <dfn>ToWebAssemblyValue</dfn>(|v|, |type|) coerces a JavaScript va
+        1. Otherwise,
+            1. Let |values| be « ».
+            1. [=list/iterate|For each=] |r| of |result|,
+            1. [=list/Append=] [=ToJSValue=](|r|) to |values|.
+            1. let |jsReturnValue| be [$CreateArrayFromList$](|values|).
+        1. Perform [=EvaluateCall=] (|promise|.\[[Resolve]],|jsReturnValue|,false)
+    1. Return UNUSED.
+</div>
+
+Note: The extra |$Call$| in the above algorithm ensures that the creation of the Promise is separated from the fullfilling of that Promise. In effect, this allows suspension of the fullfillment to occur whilst allowing the creation of the Promise itself to continue.
+
+<div algorithm>
+    The <dfn constructor for="Suspending">Suspending(|jsFun|)</dfn> constructor, when invoked, performs the following steps:
+    1. If [$IsCallable$](|jsFun|) is false, throw a {{TypeError}}.
+    1. Let |suspendingProto| be \[[%WebAssembly.Suspending.prototype%]].
+    1. Let |susp| be the result of [$OrdinaryObjectCreate$](|suspendingProto|,«\[[WrappedFunction]]»).
+    1. Assign the \[[WrappedFunction]] internal slot of |susp| to |jsFun|.
+    1. Return |susp|.
+</div>
+
+<div algorithm>
+To <dfn>create a suspending function</dfn> from a JavaScript function |func|, with type |functype| perform the following steps:
+
+    1. Assert: [$IsCallable$](|func|).
+    1. Let |stored settings| be the <a spec=HTML>incumbent settings object</a>.
+    1. Let |hostfunc| be a [=host function=] which performs the following steps when called with arguments |arguments|:
+        1. Let |realm| be |func|'s [=associated Realm=].
+        1. Let |relevant settings| be |realm|'s [=realm/settings object=].
+        1. Let |async_context| be the [=surrounding agent=]'s [=running execution context=].
+        1. Let |map| be the [=surrounding agent=]'s associated [=Execution Context Status map=].
+        1. If the entry for |async_context| in |map| is not [=active=] then trap.
+        1. [=Prepare to run script=] with |relevant settings|.
+        1. [=Prepare to run a callback=] with |stored settings|.
+        1. Let [|parameters|] → [|resultTypes|] be |functype|.
+        1. Let |jsArguments| be the result of [=coerce WebAssembly arguments=](|parameters|,|arguments|).
+        1. Let |ret| be [=?=] [$Call$](|func|, undefined, |jsArguments|).
+        1. [=Clean up after running a callback=] with |stored settings|.
+        1. [=Clean up after running script=] with |relevant settings|.
+        1. Assert: |ret|.\[[Type]] is <emu-const>throw</emu-const> or <emu-const>normal</emu-const>.
+        1. If |ret|.\[[Type]] is <emu-const>throw</emu-const>, then:
+            1. Let |type|, |payload| and |opaqueData| be the result of [=coerce a JavaScript exception|coercing the JavaScript exception=] |ret|.\[[Value]].
+            1. [=WebAssembly/Throw=] with |type|, |payload| and |opaqueData|.
+        1. Otherwise, if [=list/size=] of |ret| is 1 and [$IsPromise$](|ret|.\[[Value]][0]):
```

Comment:
The IsPromise AO returns true only for built-in JS promises. Userland promises and thenables return false. I want to confirm that this is the intended behavior, that userland promises are not supported by JSPI.

Reply:
It is intended to permit userland Promises also. What predicate should we use?

Reply:
By "userland promises" I meant thenables. If thenables are intended to be disallowed then IsPromise is the correct predicate to use.